core/src/ldml/ldml_processor.cpp

```diff
@@ -59,23 +59,130 @@ namespace {
+ * Note double parens
+ * \def KMXPLUS_DEBUG
+ */
+#if KMXPLUS_DEBUG
+#include <iostream>
+#define KMXPLUS_NOTEQUAL(expect, actual) std::cerr << __FILE__ << ":" << __LINE__ << ": ASSERT FAILED: " \
+              << #actual << "=" << (actual) << ", expected " << (expect) << std::endl
+#define KMXPLUS_PRINTLN(msg) std::cerr  << __FILE__ << ":" << __LINE__ << ": " msg << std::endl;
+#else
+#define KMXPLUS_NOTEQUAL(expect, actual)
+#define KMXPLUS_PRINTLN(msg)
+#endif
+
 namespace km {
 namespace kbp {
 
 ldml_processor::ldml_processor(path const & kb_path, const std::vector<uint8_t> &data)
 : abstract_processor(
     keyboard_attributes(kb_path.stem(), KM_KBP_LMDL_PROCESSOR_VERSION, kb_path.parent(), {})
-  ), _valid(false), rawdata(data) // TODO-LDML: parse the data, don't just copy it
+  ), _valid(false), vkey_to_string()
 {
+
+/**
+ * Assert something, just in this function
+ * If fails, print a message and exit as invalid
+ * \def KMXPLUS_ASSERT
+ */
+#define KMXPLUS_ASSERT(expect,actual) if ((expect) != (actual)) { \
+    KMXPLUS_NOTEQUAL(expect, actual); \
+    _valid = false; \
+    return; \
+  }
+  const kmx::COMP_KMXPLUS_SECT *sect = nullptr;
+  const kmx::COMP_KMXPLUS_STRS *strs = nullptr;
+  const kmx::COMP_KMXPLUS_KEYS *keys = nullptr;
+
   // TODO-LDML: load the file from the buffer (KMXPlus format)
   // Note: kb_path is essentially debug metadata here
-  assert(data.size() != 0);
+  KMXPLUS_ASSERT(true, data.size() > sizeof(kmx::COMP_KEYBOARD_EX));
 
+  // Locate the structs here, but still retain ptrs to the raw structs.
   const kmx::PCOMP_KEYBOARD comp_keyboard = (kmx::PCOMP_KEYBOARD)data.data();
+  KMXPLUS_ASSERT(true, !!(comp_keyboard->dwFlags & KF_KMXPLUS));
+  const kmx::COMP_KEYBOARD_EX* ex = reinterpret_cast<const kmx::COMP_KEYBOARD_EX*>(comp_keyboard);
+
+  // validate size and offset
+  KMXPLUS_ASSERT(true, ex->kmxplus.dwKMXPlusSize >  LDML_LENGTH_SECT);
+  KMXPLUS_ASSERT(true, ex->kmxplus.dwKMXPlusSize <  (data.size() - sizeof(kmx::COMP_KEYBOARD_EX)));
+  KMXPLUS_ASSERT(true, ex->kmxplus.dpKMXPlus     >= sizeof(kmx::COMP_KEYBOARD_EX));
+  KMXPLUS_ASSERT(true, ex->kmxplus.dpKMXPlus     <  data.size());
+
+  // calculate pointers to start and end of kmxplus
+  const uint8_t * const kmxplusdata   = data.data() + ex->kmxplus.dpKMXPlus;      // Start of + data
+  // const uint8_t* kmxpluslimit  = kmxplusdata    + ex->kmxplus.dwKMXPlusSize;  // End of   + data
+
+  // Now load sections.
+
+  // This will validate (and possibly print) all data.
+  // TODO-LDML: we will be replacing this validation with validation-as-we-go below.
+  KMXPLUS_ASSERT(true, kmx::validate_kmxplus_data(kmxplusdata));
+
+  // Get out the SECT header
+  {
+    const KMX_DWORD offset = 0;
+    sect = kmx::as_kmxplus_sect(kmxplusdata+offset);
+    KMXPLUS_ASSERT(true, sect != nullptr);
+    // Specified data size fits in total
+    KMXPLUS_ASSERT(true, (offset+sect->header.size) < ex->kmxplus.dwKMXPlusSize);
+  }
```

Comment:
For unit testing and reusability, it would be good if these could all be extracted to separate functions, in a separate .cpp. Could even make them a separate class.

I'd like to see the common components of these in a single function:
* offset tests
* size tests
* ident tests

```
KMXPLUS_ASSERT(true, (offset+sect->header.size) < ex->kmxplus.dwKMXPlusSize);
```
This should be `<=` in all cases because the section may be the last thing in the kmxplus section.

Reply:
> This is looking good -- but I think it'd be good to restructure the file loading and validation here before it solidifies.

thanks for the comments.  Separate functions make sense.  Perhaps they can be part of the kmx_plus.h structs.